### PR TITLE
Add support for generating ESLint Flat Config

### PR DESCRIPTION
**Pull request by conradhale**
Will fix https://github.com/vuejs/eslint-plugin-vue/issues/1291

Needs more work, haven't had the time to work on this lately.

I was thinking it would be better to replace the generated config files with JS functions that dynamically return the correct configuration. Should decrease the size of the library and make the code a bit cleaner with the addition of flat configs and probably wouldn't impact runtime performance much. Most other eslint plugins seem to be doing this as well.

### CODE REVIEW

**Comment by conradhale** [comment]
It might be a while before I have time to work on this, if any maintainers want to take over this PR, feel free to make any edits.

**Comment by FloEdelmann** [comment]
@conradhale Thanks! I added some more commits, could you please have a look at them? I am not sure what is still missing to have full Flat config support, so I am handing this back to you again :slightly_smiling_face:

File: lib/configs/flat/base.js
Comment (aladdin-add) [inline lib/configs/flat/base.js:18]: In most cases, you don't need to specify `parserOptions`. just use:

```js
  languageOptions: {ecmaVersion: 2020, sourceType: 'module'}
```
eslint will pass it to the parser. So you don't need to specify global.es2015 either.

**Review by aladdin-add — CHANGES_REQUESTED** [review]

File: lib/configs/flat/base.js
Comment (aladdin-add) [reply lib/configs/flat/base.js:18]: https://eslint.org/docs/next/use/configure/language-options#specifying-javascript-options

File: lib/index.js
Comment (ota-meshi) [inline lib/index.js:266]: I would like to change `flat/essential`, `flat/recommended`, and `flat/strongly-recommended` to the config for Vue3. Use the `vue2-` prefix to config for Vue2 instead.

**Comment by aladdin-add** [comment]
@ota-meshi Is there anything blocking this PR from being merged now? I have been implementing `npm init @eslint/config` for the flat configs, so it would be great see it's officially supported in the plugin. 😄

**Comment by aladdin-add** [comment]
@conradhale I can pick this up! (if you don't mind :)

**Comment by ota-meshi** [comment]
I would like the configuration to be renamed.
Correspond as follows:
- (Legacy -> Flat)
- essential -> flat/vue2-essential
- strongly-recommended -> flat/vue2-strongly-recommended
- recommended -> flat/vue2-recommended
- vue3-essential -> flat/essential
- vue3-strongly-recommended -> flat/strongly-recommended
- vue3-recommended -> flat/recommended
https://github.com/vuejs/eslint-plugin-vue/pull/2319/files#r1465649137

We also need test code to test the flat configuration.

**Pull request closed without merging**
